lib/vector-core/src/buffers/disk/leveldb_buffer/mod.rs

```diff
@@ -0,0 +1,142 @@
+mod key;
+mod reader;
+mod writer;
+
+use super::{DataDirError, Open};
+use crate::buffers::Acker;
+use bytes::Bytes;
+use futures::task::AtomicWaker;
+use key::Key;
+use leveldb::database::{
+    batch::Writebatch,
+    iterator::{Iterable, LevelDBIterator},
+    options::{Options, ReadOptions},
+    Database,
+};
+use reader::Reader;
+use snafu::ResultExt;
+use std::convert::{TryFrom, TryInto};
+use std::fmt::Debug;
+use std::{
+    collections::VecDeque,
+    marker::PhantomData,
+    path::Path,
+    sync::{atomic::AtomicUsize, Arc, Mutex},
+};
+pub use writer::Writer;
+
+/// How much of disk buffer needs to be deleted before we trigger compaction.
+const MAX_UNCOMPACTED_DENOMINATOR: usize = 10;
+
+#[derive(Default)]
+pub struct Buffer<T> {
+    phantom: PhantomData<T>,
+}
+
+/// Read the byte size of the database
+///
+/// There is a mismatch between leveldb's mechanism and vector's. While vector
+/// would prefer to keep as little in-memory as possible leveldb, being a
+/// database, has the opposite consideration. As such it may mmap 1000 of its
+/// LDB files into vector's address space at a time with no ability for us to
+/// change this number. See [leveldb issue
+/// 866](https://github.com/google/leveldb/issues/866). Because we do need to
+/// know the byte size of our store we are forced to iterate through all the LDB
+/// files on disk, meaning we impose a huge memory burden on our end users right
+/// at the jump in conditions where the disk buffer has filled up. This'll OOM
+/// vector, meaning we're trapped in a catch 22.
+///
+/// This function does not solve the problem -- leveldb will still map 1000
+/// files if it wants -- but we at least avoid forcing this to happen at the
+/// start of vector.
+fn db_initial_size(path: &Path) -> Result<usize, DataDirError> {
+    let mut options = Options::new();
+    options.create_if_missing = true;
+    let db: Database<Key> = Database::open(&path, options).with_context(|| Open {
+        data_dir: path.parent().expect("always a parent"),
+    })?;
+    Ok(db.value_iter(ReadOptions::new()).map(|v| v.len()).sum())
+}
+
+impl<T> Buffer<T>
+where
+    T: Send + Sync + Unpin + TryInto<Bytes> + TryFrom<Bytes>,
+    <T as TryInto<bytes::Bytes>>::Error: Debug,
+    <T as std::convert::TryFrom<bytes::Bytes>>::Error: Debug,
```

Comment:
```suggestion
    <T as std::convert::TryFrom<bytes::Bytes>>::Error: Debug + Display,
```

Reply:
`Display` wasn't needed here to compile and when feasible I prefer to leave type constraints lax.

Reply:
Yes, but `to leave type constraints lax` fair enough. Let it stay like that.